rfcs/20221124-monorepo.md

```diff
@@ -0,0 +1,446 @@
+
+1. GitHub Milestones:
+
+   Milestones are confined by repository boundaries. To have milestones that
+   cover issues in different layers of our stack, the only way is for them to be
+   in a monorepo. This is a limitation imposed by GitHub and there is no
+   workaround for this.
+
+The overarching theme is that there are workarounds for everything. We have been
+working with split repos quite productively for over a year. My proposition is
+the the monorepo solutions are better than workarounds.
+
+The [integration](#step-4-integration) section in the latter part of the
+document describes more interesting outcomes made possible by the monorepo. They
+may or may not be exclusive to monorepos but they're surely made easier by it.
+
+## Migration path
+
+First we will merge the API and the frontend repos into `WordPress/openverse`.
+This decision was made for the following reasons.
+
+1. API and frontend are tightly linked. The frontend is a direct consumer of
+   what the API produces.
+
+1. The API and frontend form the "service" side of Openverse that directly faces
+   the users (both API consumers and Search engine users).
+
+1. The frontend uses ECS deployments and the API is well on the same track. This
+   makes it possible for them to share some deployment code.
+
+1. I am very familiar with the scripts repo, the API and the frontend so merging
+   them would be easier. Adding a third component would make the task daunting.
+
+1. Merging incurs a productivity hit for the initial transition. So merging
+   everything in one swoop is not ideal. While we merge these three, effort can
+   be diverted to the catalog.
+
+1. The API’s comprehensive tooling for developer documentation can benefit
+   frontend devs and create a unified docs site for contributors.
+
+1. The merge of two JavaScript codebases provides fertile ground for testing
+   `pnpm` workspaces.
+
+   - It also allows us to merge the browser extension later and split the design
+     system/component library stuff into a separate package.
+
+1. The API is already organised by stack folders so the `frontend/` directory
+   will fit right in with the others like `api/` and `ingestion_server/`.
+   Similarly the scripts repo is nicely organised in folders, reducing
+   conflicts.
+
+1. The API and frontend share identical tooling for Git hooks, linting and
+   formatting. We will fight our tools less and encounter minimal friction.
+
+   - The frontend's approach for `pre-commit` expanded this type of usage to the
+     API as well!
+
+   - We're expanding the use of double-quoted strings to JavaScript to further
+     unify our style guides.
+
+1. The entire system can be integration tested during releases. The real API,
+   populated with test data, can replace the Talkback server as long as we
+   disable network calls and make output deterministic.
+
+The `WordPress/openverse` repo will absorb the `WordPress/openverse-api` and
```

Comment:
Another great benefit to this that we've mentioned previously is that `WordPress/openverse` is a great, simple name and logical entrypoint to the code for our users.